BLSQ/iaso
# WC2-171 Entities Deduplication Feature

Give a way to the users to find duplicated entities (per same entity type and account)
Allow them to run algorithms/analyzes that will find the duplicates (by choosing which fields on the reference form of the entity type to analyze + other params for the algorithm)
List the found duplicates and allows the users to ignore or merge them interactively.


Related JIRA tickets : WC2-171, WC2-137, WC2-230, WC2-231, WC2-236

## Self proofreading checklist


## Changes

Backend + Background tasks and frontend for the feature.
One implementation of an algorithm to find duplicates.

## How to test

In order to test locally you will need to have at least one Entity Type setup correctly as well as multiple entities with the same or similar attributes (reference form of the Entity Type) filled. You can find in `iaso/tests/fixtures/test_form_deduplication.xlsx` a simple form with 3 fields : Prenom, Nom, Age that you can use (see the test code in `iaso/tests/api/deduplication/test_entities_deduplication.py`

After that you need to create an initial analyze configuration using the API, you need to send a POST request to the URL `/api/entityduplicates_analyzes/` with a JSON payload similar to this one (depending on the fields you have in your reference form) :

```json
{
  "fields": [
    "What_is_the_father_s_name",
    "What_is_the_child_s_name"
  ],
  "parameters": {},
  "entity_type_id": "2"
}
```

If you are in local, You need to have a background queue job worker running and it can be launched using `./manage.py tasks_worker` or `docker-compose exec iaso ./manage.py tasks_worker` if you are in docker.

After that if there are similar entities, entity duplicates will be created and you can use the frontend to see and ignore/merge them use the dashboard menu Beneficiaries > Duplicates or visit this url : `/dashboard/entities/duplicates/accountId/2/order/id/pageSize/20/page/1`


## Print screen / video


<img width="1169" alt="Screenshot 2023-06-09 at 12 19 02" src="https://github.com/BLSQ/iaso/assets/383344/573371fd-391d-4ee8-9143-134a11751105">
<img width="1166" alt="Screenshot 2023-06-09 at 12 16 00" src="https://github.com/BLSQ/iaso/assets/383344/c27991b8-d4cb-4807-b105-2d4611b3353f">
<img width="804" alt="Screenshot 2023-06-09 at 12 09 28" src="https://github.com/BLSQ/iaso/assets/383344/b6e0a00d-c564-4fbb-87c1-a055cd290cc9">
